libnet: Endpoint: remove isAnonymous & myAliases

akerouanton opened the pull request:
**- What I did**

The `Endpoint` struct was containing two confusing fields:

1. `isAnonymous`: was set on endpoints which name shouldn't be taken into account when inserting DNS records into libnetwork's `Controller.svcRecords` (and into the NetworkDB). However, in that case the endpoint's aliases would still be used to create DNS records; thus, making those "anonymous endpoints" not so anonymous.
2. `myAliases`: there's also an `aliases` field in the `Endpoint` struct. `myAliases` was storing extra DNS names for the endpoint, while `aliases` stores extra DNS names for _other_ endpoints (but those extra DNS names are only accessible for the endpoint storing them -- yeah, it's confusing too).

To make it worse, the code that built the list of DNS names for a given endpoint was split in two:

1. `updateNetworkConfig` (in the `daemon` pkg) was adding the endpoint short ID and hostname to the list of aliases ;
2. In libnetwork, to handle "anonymous" endpoints as described above.

This PR replaces both fields with a new `DNSNames` containing _all_ (non-fully qualified) DNS names for a given endpoint. It's now built by the daemon during `updateNetworkConfig` and reimplements the same logic for aliases as before (a test has been added to make sure of that). An "anonymous" endpoint is now an endpoint with no `DNSNames`.

This new field is also stored in the "operational data" (ie. state) of the `EndpointSettings` struct exposed by the API via `GET /containers/{name:.*}/json`, making it easier for end-users and scripts to know every DNS names a container has on a specific network.

**- How to verify it**

`docker run --network testnet` with a variety of:

* `--name` ;
* `--hostname` ;
* `--network-alias` ;

**- A picture of a cute animal (not mandatory but encouraged)**

![](https://www.nhm.ac.uk/resources/visit/wpy/2018/large/webp/14.webp)

akerouanton committed:
daemon: endpoints on default nw aren't anonymous

They just happen to exist on a network that doesn't support DNS-based
service discovery (ie. no embedded DNS servers are started for them).

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
daemon: no more IsAnonymousEndpoint

The semantics of an "anonymous" endpoint has always been weird: it was
set on endpoints which name shouldn't be taken into account when
inserting DNS records into libnetwork's `Controller.svcRecords` (and
into the NetworkDB). However, in that case the endpoint's aliases would
still be used to create DNS records; thus, making those "anonymous
endpoints" not so anonymous.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
daemon: build the list of endpoint's DNS names

Instead of special-casing anonymous endpoints in libnetwork, let the
daemon specify what (non fully qualified) DNS names should be associated
to container's endpoints.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet: Use Endpoint.dnsNames to create DNS records

Instead of special-casing anonymous endpoints, use the list of DNS names
associated to the endpoint.

`(*Endpoint).isAnonymous()` has no more uses, so let's delete it.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet: add a new dnsNames property to Endpoint

This new property is meant to replace myAliases and anonymous
properties.

The end goal is to get rid of both properties by letting the daemon
determine what (non fully qualified) DNS names should be associated to
them.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet: remove Endpoint.myAliases

This property is now unused, let's get rid of it.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet: update dnsNames on ContainerRename

The `(*Endpoint).rename()` method is changed to only mutate `ep.name`
and let a new method `(*Endpoint).UpdateDNSNames()` handle DNS updates.

As a consequence, the rollback code that was part of
`(*Endpoint).rename()` is now removed, and DNS updates are now
rolled back by `ContainerRename`.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

robmry reviewed (approved):
LGTM

corhere commented on daemon/rename.go:
```diff
@@ -118,10 +119,68 @@ func (daemon *Daemon) ContainerRename(oldName, newName string) (retErr error) {
+			epConfig *network.EndpointSettings
+			dnsNames []string
+		}
+		defer func() {
+			if retErr != nil {
+				for _, update := range updatedEps {
+					update.epConfig.DNSNames = update.dnsNames
+					if err := update.ep.UpdateDNSNames(update.epConfig.DNSNames); err != nil {
+						log.G(context.TODO()).WithFields(log.Fields{
+							"sandboxID": sid,
+							"oldName":   oldName,
+							"newName":   newName,
+							"error":     err,
+						}).Errorf("failed to revert DNSNames update")
+					}
+				}
+			}
+		}()
```
Instead of managing the undo stack by hand, why not let the compiler handle it all for you by lowering the `defer` inside the loop?

corhere reviewed (changes requested):
How can the `anonymous` and `myAliases` fields be removed from the endpoint KVObjects without migration code? Wouldn't DNS resolution be subtly broken on containers carried across a daemon upgrade?

akerouanton committed:
libnet: populate Endpoint.dnsNames on UnmarshalJSON

This new property will be empty if the daemon was upgraded with
live-restore enabled. To not break DNS resolutions for restored
containers, we need to populate dnsNames based on endpoint's myAliases &
anonymous properties.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet: document what Network.networkType represents

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton commented:
> How can the anonymous and myAliases fields be removed from the endpoint KVObjects without migration code? Wouldn't DNS resolution be subtly broken on containers carried across a daemon upgrade?

Yeah, right. I overlooked this part. I added a commit to manage this case (see https://github.com/moby/moby/pull/46853/commits/de2303b32dde9950d95e35697354180b90bb110d). However, I now realize the field `EndpointSettings.DNSNames` should be set on restore too. I also resolved the other comments.

Some integration tests are broken as they expect the container short ID to be in `EndpointSettings.Aliases`. I think it's not a backward-incompatible change but what's your opinion @thaJeztah @corhere? At least, I should update `version-history.md` to make it clear.

corhere commented on libnetwork/endpoint.go:
```diff
@@ -64,9 +67,8 @@ func (ep *Endpoint) MarshalJSON() ([]byte, error) {
 	}
 	epMap["sandbox"] = ep.sandboxID
-	epMap["anonymous"] = ep.anonymous
+	epMap["dnsNames"] = ep.dnsNames
 	epMap["disableResolution"] = ep.disableResolution
-	epMap["myAliases"] = ep.myAliases
```
I wonder if we should keep populating `anonymous` and `myAliases` in the marshaled objects so users could downgrade. Though only live-restored containers would be affected given that it's the endpoint object. As we already document that [live-restore is only supported across patch upgrades](https://docs.docker.com/config/containers/live-restore/#live-restore-during-upgrades) we have no obligation to support major-version _downgrades_.

akerouanton replied:
Since we explicitly specify we don't support live-restore for major upgrades, I think we shouldn't care much about live-restore for major downgrades.

corhere reviewed (commented):
> Some integration tests are broken as they expect the container short ID to be in `EndpointSettings.Aliases`. I think it's not a backward-incompatible change but what's your opinion @thaJeztah @corhere?

It's an API _configuration_ field and integration tests are explicitly testing for that behaviour. [Someone's workflow is going to break.](https://www.hyrumslaw.com/) I don't think it's safe to change the semantics of `EndpointSettings.Aliases` on released API versions.

akerouanton committed:
libnet: remove Endpoint.anonymous

No more concept of "anonymous endpoints". The equivalent is now an
endpoint with no DNSNames set.

Some of the code removed by this commit was mutating user-supplied
endpoint's Aliases to add container's short ID to that list. In order to
preserve backward compatibility for the ContainerInspect endpoint, this
commit also takes care of adding that short ID (and the container
hostname) to `EndpointSettings.Aliases` before returning the response.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

corhere commented on libnetwork/endpoint.go:
```diff
@@ -1178,3 +1159,39 @@ func (c *Controller) cleanupLocalEndpoints() error {
+// BuildEndpointDNSNames constructs the list of DNSNames that should be assigned to a given endpoint. Any zero-value
+// parameter will be skipped. The order within the returned slice is important as the first entry will be used to
+// generate the PTR records (for IPv4 and v6) associated to this endpoint.
+func BuildEndpointDNSNames(name, hostname, id string, aliases []string) []string {
```
This function seems much more related to the daemon business logic than libnetwork: it has specific knowledge about how to transform container names and IDs into appropriate DNS names. And with one exception, it is exclusively used by the daemon to transform container configuration into libnetwork endpoint config. I think this function belongs as an unexported function in the `daemon` package. The one exception, the migration code, only uses a subset of what this function does. The migration code within libnetwork could simply be open-coded. `dedupSlice` could live in some `internal` utility package for use in both. Or we could double down on "live-restore across major upgrades is unsupported" and not bother migrating endpoints at all.

corhere commented on libnetwork/endpoint.go:
```diff
@@ -192,7 +195,25 @@ func (ep *Endpoint) UnmarshalJSON(b []byte) (err error) {
 	json.Unmarshal(ma, &myAliases) //nolint:errcheck
-	ep.myAliases = myAliases
+
+	_, hasDNSNames := epMap["dnsNames"]
+	dn, _ := json.Marshal(epMap["dnsNames"])
+	var dnsNames []string
+	json.Unmarshal(dn, &dnsNames)
+	ep.dnsNames = dnsNames
+
+	if !hasDNSNames {
+		// The field dnsNames was introduced in v25.0. If we don't have it, the on-disk state was written by an older
+		// daemon, thus we need to populate dnsNames based off of myAliases and anonymous values.
+		var name string
+		if !anonymous {
+			name = ep.name
+		}
+		// Hostname and id fields are purposefully left empty here as they should have been put into myAliases when the
+		// endpoint was originally created.
+		ep.dnsNames = BuildEndpointDNSNames(name, "", "", myAliases)
+	}
```
```suggestion
	if dnsNames, ok := epMap["dnsNames"]; ok {
		dn, _ := json.Marshal(dnsNames)
		json.Unmarshal(dn, &ep.dnsNames)
	} else {
		// The field dnsNames was introduced in v25.0. If we don't have it, the on-disk state was written by an older
		// daemon, thus we need to populate dnsNames based off of myAliases and anonymous values.
		if !anonymous {
			myAliases = append([]string{name}, myAliases...)
		}
		ep.dnsNames = sliceutil.Dedup(myAliases) // dedupSlice, but in an internal/ package that does not yet exist.
	}
```

corhere commented on libnetwork/agent.go:
```diff
@@ -628,9 +628,15 @@ func (ep *Endpoint) addServiceInfoToCluster(sb *Sandbox) error {
+	copy(dnsNames, ep.dnsNames)
+	ep.mu.Unlock()
+
+	primaryDNSName := dnsNames[0]
+	var dnsAliases []string
+	if len(dnsNames) > 1 {
+		dnsAliases = dnsNames[1:]
 	}
```
```suggestion
	primaryDNSName, dnsAliases := dnsNames[0], dnsNames[1:]
```
for consistency with line 713

corhere commented on libnetwork/network.go:
```diff
@@ -1299,40 +1299,34 @@ func (n *Network) updateSvcRecord(ep *Endpoint, isAdd bool) {
-				n.deleteSvcRecords(ep.ID(), myAliases[0], serviceID, iface.Address().IP, ipv6, true, "updateSvcRecord")
+		for i, dnsName := range dnsNames {
+			// ipMapUpdate indicates whether PTR records should be updated.
+			var ipMapUpdate bool
+			if i == 0 {
+				ipMapUpdate = true
 			}
```
```suggestion
			ipMapUpdate := i == 0
```

thaJeztah commented:
Something broke;

```
daemon/container_operations.go:681:31: undefined: stringid
85.20 daemon/rename.go:148:57: cannot use epConfig (variable of type *"github.com/docker/docker/daemon/network".EndpointSettings) as *"github.com/docker/docker/api/types/network".EndpointSettings value in argument to buildEndpointDNSNames

thaJeztah commented on libnetwork/endpoint.go:
```diff
@@ -586,71 +586,52 @@ func (ep *Endpoint) sbJoin(sb *Sandbox, options ...EndpointOption) (err error) {
-				n.updateSvcRecord(ep, true)
-			}
-		}()
+		nw.updateSvcRecord(ep, false)
+
+		ep.dnsNames = dnsNames
+		nw.updateSvcRecord(ep, true)
```
That `updateSvcRecord` bool (add / remove) is really awkward 😞 (we should consider having `addSvcRecord` / `deleteSvcRecord` functions (or similar)

akerouanton replied:
I agree, although we already have methods named `addSvcRecords` and `deleteSvcRecords` (those ones are even getting called by `updateSvcRecord`). The notion of "svc records" is also quite awkward. We should have methods that better model the problem space and uses the DNS terminology (cf. my TODO about PTR records).

thaJeztah replied:
> we already have methods named addSvcRecords and deleteSvcRecords (those ones are even getting called by updateSvcRecord).

Yes, I noticed those! And yes, it's really just an observation (that made me weep). I'm not always a fan of abstracting things, but for this case, I'd definitely be "+1" on adding `delete` and `add` wrappers instead of a magic "add/remove" boolean.

thaJeztah commented on daemon/inspect.go:
```diff
@@ -27,6 +28,23 @@ func (daemon *Daemon) ContainerInspect(ctx context.Context, name string, size bo
 		return daemon.containerInspectPre120(ctx, name)
 	case versions.Equal(version, "1.20"):
 		return daemon.containerInspect120(name)
+	case versions.LessThan(version, "1.44"):
+		ctr, err := daemon.ContainerInspectCurrent(ctx, name, size)
+		if err != nil {
+			return nil, err
+		}
+
+		shortCID := stringid.TruncateID(ctr.ID)
+		for _, ep := range ctr.NetworkSettings.Networks {
+			ep.Aliases = append(ep.Aliases, shortCID)
+			// Container hostname defaults to the container short ID if no hostname was specified -- we don't want
+			// duplicates.
+			if ctr.Config.Hostname != shortCID {
+				ep.Aliases = append(ep.Aliases, ctr.Config.Hostname)
+			}
+		}
```
Double-checking here; this is adapting the response for API 1.29 - 1.43.  Is the container short-id to be found elsewhere in API 1.44? Are we still supporting that to be resolved (or was this only used as hostname _inside_ the container? i.e. related to the logic that was added specifically for that; https://github.com/moby/moby/blob/74cf9a0f60d5b91a36d3be3f9c0e3790cdd0e24a/pkg/stringid/stringid.go#L54-L59


- https://github.com/moby/moby/commit/9b7054fd0bc707f7f8e5c9ed9ada5a1664a19a34 /  https://github.com/moby/moby/pull/3881
- https://github.com/moby/moby/issues/3869

akerouanton replied:
> Is the container short-id to be found elsewhere in API 1.44?

I'm not sure why you're asking this question. I didn't check, but I don't see how any other short CID fields would relate to that change.

> Are we still supporting that to be resolved (or was this only used as hostname inside the container? i.e. related to the logic that was added specifically for that;

Yeah, the short CID is still a valid DNS name -- it can be resolved. So the logic in `stringid.GenerateRandomID()` is still relevant.

thaJeztah replied:
> I'm not sure why you're asking this question. I didn't check, but I don't see how any other short CID fields would relate to that change.

Ah, yes, sorry, I honestly got lost in the weeds a bit with all the fields (and deprecated ones, and replacements for them).

I was wondering if there was a replacement for the information we're removing. In most (if not all) cases where we moved fields, the new ones were still providing the same information, and I wasn't sure if that was the case here (nor if there were use-cases for it).

thaJeztah replied:
Right, so testing with this PR;

- in API 1.28 - 1.43, the `NetworkSettings.Networks["foo'}.Aliases` always contains the container's short-ID as alias
- in API 1.44, we're adding the new `DNSNames` field, but _ALSO_ removing the container's short-ID from the `NetworkSettings.Networks["foo'}.Aliases` field.

See:


```bash
docker network create mynetwork

docker run -d --name mycontainer1 --network mynetwork nginx:alpine
docker run -d --name mycontainer2 --network-alias myalias --network mynetwork nginx:alpine
```

### On API 1.43

```bash
curl -s --unix-socket /var/run/docker.sock http://localhost/v1.43/containers/mycontainer1/json | jq .NetworkSettings.Networks
{
  "mynetwork": {
    "IPAMConfig": null,
    "Links": null,
    "Aliases": [
      "2ba467ea502f"
    ],
    "MacAddress": "02:42:ac:13:00:02",
    "NetworkID": "68bea4575d8ee10a16151e90f58786f4afd62135fb5a2b412e16140b0ced48a5",
    "EndpointID": "e942106dbc637a66e770a8bd5bb79f61421ec3b8485ea2f1c186116427b9abc9",
    "Gateway": "172.19.0.1",
    "IPAddress": "172.19.0.2",
    "IPPrefixLen": 16,
    "IPv6Gateway": "",
    "GlobalIPv6Address": "",
    "GlobalIPv6PrefixLen": 0,
    "DriverOpts": null,
    "DNSNames": [
      "mycontainer1",
      "2ba467ea502f"
    ]
  }
}
curl -s --unix-socket /var/run/docker.sock http://localhost/v1.43/containers/mycontainer2/json | jq .NetworkSettings.Networks
{
  "mynetwork": {
    "IPAMConfig": null,
    "Links": null,
    "Aliases": [
      "myalias",
      "632ae41df334"
    ],
    "MacAddress": "02:42:ac:13:00:03",
    "NetworkID": "68bea4575d8ee10a16151e90f58786f4afd62135fb5a2b412e16140b0ced48a5",
    "EndpointID": "a0f4c5879225870e8c054a5af1c8e892e57a29557999dfbe9a577ac4eb988c2d",
    "Gateway": "172.19.0.1",
    "IPAddress": "172.19.0.3",
    "IPPrefixLen": 16,
    "IPv6Gateway": "",
    "GlobalIPv6Address": "",
    "GlobalIPv6PrefixLen": 0,
    "DriverOpts": null,
    "DNSNames": [
      "mycontainer2",
      "myalias",
      "632ae41df334"
    ]
  }
}
```

### On API 1.44

```
curl -s --unix-socket /var/run/docker.sock http://localhost/v1.44/containers/mycontainer1/json | jq .NetworkSettings.Networks
{
  "mynetwork": {
    "IPAMConfig": null,
    "Links": null,
    "Aliases": null,
    "MacAddress": "02:42:ac:13:00:02",
    "NetworkID": "68bea4575d8ee10a16151e90f58786f4afd62135fb5a2b412e16140b0ced48a5",
    "EndpointID": "e942106dbc637a66e770a8bd5bb79f61421ec3b8485ea2f1c186116427b9abc9",
    "Gateway": "172.19.0.1",
    "IPAddress": "172.19.0.2",
    "IPPrefixLen": 16,
    "IPv6Gateway": "",
    "GlobalIPv6Address": "",
    "GlobalIPv6PrefixLen": 0,
    "DriverOpts": null,
    "DNSNames": [
      "mycontainer1",
      "2ba467ea502f"
    ]
  }
}
curl -s --unix-socket /var/run/docker.sock http://localhost/v1.44/containers/mycontainer2/json | jq .NetworkSettings.Networks
{
  "mynetwork": {
    "IPAMConfig": null,
    "Links": null,
    "Aliases": [
      "myalias"
    ],
    "MacAddress": "02:42:ac:13:00:03",
    "NetworkID": "68bea4575d8ee10a16151e90f58786f4afd62135fb5a2b412e16140b0ced48a5",
    "EndpointID": "a0f4c5879225870e8c054a5af1c8e892e57a29557999dfbe9a577ac4eb988c2d",
    "Gateway": "172.19.0.1",
    "IPAddress": "172.19.0.3",
    "IPPrefixLen": 16,
    "IPv6Gateway": "",
    "GlobalIPv6Address": "",
    "GlobalIPv6PrefixLen": 0,
    "DriverOpts": null,
    "DNSNames": [
      "mycontainer2",
      "myalias",
      "632ae41df334"
    ]
  }
}
```

While that's not a change in the API _fields_, it _is_ a change in the meaning of an API field, which likely means we 

- must _deprecate_ that functionality (`Aliases` also contains the container's short-ID as network-scoped alias)
- keep the behavior for API 1.44 (which would return the short-ID _BOTH_ in `Aliases` _AND_ in `DNSNames`
- remove it for API 1.45

thaJeztah replied:
Given that API 1.44 and below would _always_ have at least one entry in `Aliases` (it would never be `nil`?), perhaps we should also make sure that it always returns a slice (empty `[]` if there's no aliases) instead of `nil` 🤔

akerouanton commented on libnetwork/agent.go:
```diff
@@ -628,10 +628,12 @@ func (ep *Endpoint) addServiceInfoToCluster(sb *Sandbox) error {
+	ep.mu.Lock()
+	dnsNames := make([]string, len(ep.dnsNames))
+	copy(dnsNames, ep.dnsNames)
+	ep.mu.Unlock()
```
I tried to remove some of them by adding a method `ep.getDNSNames()` as you suggested, but there're still some `copy()` hanging around as they're not on the same type.

akerouanton commented:
I investigated the broken docker-py test to discover that the `Container.run()` method of docker-py returns the container struct as it was after the `Container.create()` call, not after the container was actually started.

Because of that, the failing integration test `test_run_with_networking_config` expects only `Aliases` that were submitted during the ContainerCreate call. Before this change, the container short ID was included in the `Aliases` field at run-time. However, to provide backward-compatibility this change always includes it. This python test will have to be fixed to not make a strict equality check, but instead check that the expected Aliases are there -- and ignore the additionnal Aliases.

thaJeztah reviewed (approved):
LGTM
